sway-core/src/semantic_analysis/context.rs

```diff
@@ -0,0 +1,186 @@
+use crate::{
+    parse_tree::declaration::Purity,
+    semantic_analysis::{
+        ast_node::Mode,
+        declaration::{EnforceTypeArguments, MonomorphizeHelper},
+        Namespace,
+    },
+    type_engine::{insert_type, unify_with_self, TypeId, TypeInfo},
+    CompileResult, CompileWarning, TypeArgument, TypeError,
+};
+use sway_types::{span::Span, Spanned};
+
+/// Contextual state tracked and accumulated throughout type-checking.
+pub struct Context<'ns> {
+    pub(crate) namespace: &'ns mut Namespace,
+
+    // The following set of fields are intentionally private. When a `Context` is passed into a new
+    // node during type checking, these fields should be updated using the `with_*` methods which
+    // provides a new `Context`, ensuring we don't leak our changes into the parent nodes.
+    self_type: TypeId,
+    /// While type-checking an expression, this indicates the expected type.
+    ///
+    /// Assists type inference.
+    type_annotation: TypeId,
+    mode: Mode,
+    // TODO: Should this be removed to `CompileError` and `CompileWarning`?
+    help_text: &'static str,
+    /// Tracks the purity of the context, e.g. whether or not we should be allowed to write to
+    /// storage.
+    purity: Purity,
+}
+
+impl<'ns> Context<'ns> {
+    /// Initialise a context at the top-level of a module with its namespace.
+    ///
+    /// Initializes with:
+    ///
+    /// - type_annotation: unknown
+    /// - mode: NoneAbi
+    /// - help_text: ""
+    /// - purity: Pure
+    pub fn from_module_namespace(namespace: &'ns mut Namespace) -> Self {
+        Self {
+            namespace,
+            type_annotation: insert_type(TypeInfo::Unknown),
+            help_text: "",
+            // TODO: Contract? Should this be passed in based on program kind (aka TreeType)?
+            self_type: insert_type(TypeInfo::Contract),
+            mode: Mode::NonAbi,
+            purity: Purity::default(),
+        }
+    }
+
+    /// Create a new context that mutably borrows the inner `namespace` with a lifetime bound by
+    /// `self`.
+    ///
+    /// This is particularly useful when type-checking a node that has more than one child node
+    /// (very often the case). By taking the context with the namespace lifetime bound to `self`
+    /// rather than the original namespace reference, we instead restrict the returned context to
+    /// the local scope and avoid consuming the original context when providing context to the
+    /// first visited child node.
+    pub fn by_ref(&mut self) -> Context {
```

Comment:
Would you consider renaming this to `by_ref_mut`? To stay consistent with the pseudo-equivalent `ref mut` Rust pattern.

Reply:
In this case, I was aiming to follow the precedent set by [`Iterator::by_ref`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.by_ref) which does something kiiiind of similar. There can be no immutable case here, so I'm hoping there should be no ambiguity anyways.

Reply:
Ah gotcha. Feel free to ignore this then.